---
title: "eQTL_analysis"
author: "Uzezi Okinedo"
date: "1/4/2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r load required packages}
# setwd("~/helianthus_project/Helianthus_argophyllus project/")

# load tidyverse package
#library(tidyverse)
library(dplyr)
library(tidyr)
library(MatrixEQTL)
library(trio)
library(broom)
library(Biobase)
library(qvalue)

```

```{r load data}

setwd("./")

# Process SNPs for eQTL analysis
# snps1 <- read.table("Helianthus_argophyllus project/filtered_vcf.012",sep="\t",
#                   header=T,row.names=NULL)
# snps <- data.frame(t(snps1))

# Edit snp dataframe to reflect same number of rows as loci dataset
# snps <- snps[-c(1, 106133, 106134), ]
# snp_colname <- read.table("Helianthus_argophyllus project/filtered_vcf.012.indv", sep = "\t", header = F, skip = 1)
# snp_loc <- read.table("Helianthus_argophyllus project/filtered_vcf.012.pos", sep = "\t", header = F, skip = 1)
# snp_loc <- snp_loc[-c(1),]
# row.names(snp_loc) <- make.names(snp_loc$V1, unique = TRUE)
# colnames(snps) <- snp_colname$V1
# row.names(snps) <- row.names(snp_loc)
# snps <- snps[ , order(names(snps))]

# Remove genebank accessions from snps dataset
# snps$Ames449.Aligned.sortedByCoord.out.bam <- NULL
# snps$Ames695.Aligned.sortedByCoord.out.bam <- NULL
# snps$ARG1805.Aligned.sortedByCoord.out.bam <- NULL
# snps$ARG1820.Aligned.sortedByCoord.out.bam <- NULL
# snps$ARG1834.Aligned.sortedByCoord.out.bam <- NULL

# Remove missing genotypes
# snps <- subset(snps, !rowSums(snps < 0))
# k <- setdiff(row.names(snp_loc), row.names(snps))
# snp_loc <- snp_loc[!row.names(snp_loc) %in% k,]

# Compare snp data set to expression data set
expr <- read.table("normalized_readcounts.txt",sep="\t",
                   header=T,row.names=1)

# Remove missing locus
gene_loc <- read.table("gene_loc.txt", header = TRUE)
# h <- setdiff(row.names(gene_loc), row.names(expr))
# gene_loc <- gene_loc[!row.names(gene_loc) %in% h,]


# Remove btm9 from expression dataset
expr$btm9.4 <- NULL
colnames(snps) <- colnames(expr)

# Write snps, loci, and datasets into file
# write.table(snps, file = "Helianthus_argophyllus project/eQTL_snps.txt", quote = FALSE)
# write.table(snp_loc, file = "Helianthus_argophyllus project/snp_loc.txt", quote = FALSE)
# write.table(expr, file = "Helianthus_argophyllus project/normalized_readcounts.txt", quote = FALSE)
# write.table(gene_loc, file = "Helianthus_argophyllus project/gene_loc.txt", quote = FALSE)


## Location of the package with the data files.
# base.dir = find.package('MatrixEQTL');
base.dir = './';

## Settings

# Linear model to use, modelANOVA, modelLINEAR, or modelLINEAR_CROSS
useModel = MatrixEQTL::modelLINEAR; # modelANOVA, modelLINEAR, or modelLINEAR_CROSS

# Genotype file name
SNP_file_name = paste(base.dir, "eQTL_snps.txt", sep="");
snps_location_file_name = paste(base.dir, "snp_loc.txt", sep="");

# Gene expression file name
expression_file_name = paste(base.dir, "normalized_readcounts.txt", sep="");
gene_location_file_name = paste(base.dir, "gene_loc.txt", sep="");

# Covariates file name
# Set to character() for no covariates
covariates_file_name = paste(base.dir, "covariates.txt", sep="");

# Output file name
output_file_name_cis = tempfile();
output_file_name_tra = tempfile();

# Only associations significant at this level will be saved
pvOutputThreshold_cis = 2e-3;
pvOutputThreshold_tra = 1e-3;

# Error covariance matrix
# Set to numeric() for identity.
errorCovariance = numeric();
# errorCovariance = read.table("Sample_Data/errorCovariance.txt");

# Distance for local gene-SNP pairs
cisDist = 1e6;

## Load genotype data

snps = MatrixEQTL::SlicedData$new();
snps$fileDelimiter = "\t";      # the TAB character
snps$fileOmitCharacters = "NA"; # denote missing values;
snps$fileSkipRows = 1;          # one row of column labels
snps$fileSkipColumns = 1;       # one column of row labels
snps$fileSliceSize = 2000;      # read file in slices of 2,000 rows
snps$LoadFile(SNP_file_name);


## Load gene expression data

gene = MatrixEQTL::SlicedData$new();
gene$fileDelimiter = "\t";      # the TAB character
gene$fileOmitCharacters = "NA"; # denote missing values;
gene$fileSkipRows = 1;          # one row of column labels
gene$fileSkipColumns = 1;       # one column of row labels
gene$fileSliceSize = 2000;      # read file in slices of 2,000 rows
gene$LoadFile(expression_file_name);

## Load covariates

cvrt = MatrixEQTL::SlicedData$new();
cvrt$fileDelimiter = "\t";      # the TAB character
cvrt$fileOmitCharacters = "NA"; # denote missing values;
cvrt$fileSkipRows = 1;          # one row of column labels
cvrt$fileSkipColumns = 1;       # one column of row labels
if(length(covariates_file_name)>0) {
  cvrt$LoadFile(covariates_file_name);
 
   
# Remove snps with maf  
maf.list = vector('list', length(snps))
for(sl in 1:length(snps)) {
  slice = snps[[sl]];
  maf.list[[sl]] = rowMeans(slice,na.rm=TRUE)/2;
  maf.list[[sl]] = pmin(maf.list[[sl]],1-maf.list[[sl]]);
}
maf = unlist(maf.list)

hist(maf)

## Look at the distribution of MAF
hist(maf[maf<0.1],seq(0,0.1,length.out=100))

cat('SNPs before filtering:',nrow(snps))
snps$RowReorderSimple(maf>0.1);
snps$RowReorder(maf>0.1);
cat('SNPs after filtering:',nrow(snps))

}

# Remove outliers
# for( sl in 1:length(gene) ) {
# mat = gene[[sl]];
# mat = t(apply(mat, 1, rank, ties.method = "average"));
# mat = qnorm(mat / (ncol(gene)+1));
#gene[[sl]] = mat;
#}
#rm(sl, mat);

## Run the analysis
snpspos = read.table(snps_location_file_name, header = TRUE, stringsAsFactors = FALSE);
genepos = read.table(gene_location_file_name, header = TRUE, stringsAsFactors = FALSE);

me = MatrixEQTL::Matrix_eQTL_main(
  snps = snps,
  gene = gene,
  cvrt = cvrt,
  output_file_name     = output_file_name_tra,
  pvOutputThreshold     = pvOutputThreshold_tra,
  useModel = useModel,
  errorCovariance = errorCovariance,
  verbose = TRUE,
  output_file_name.cis = output_file_name_cis,
  pvOutputThreshold.cis = pvOutputThreshold_cis,
  snpspos = snpspos,
  genepos = genepos,
  cisDist = cisDist,
  pvalue.hist = "qqplot",
  min.pv.by.genesnp = FALSE,
  noFDRsaveMemory = FALSE);

unlink(output_file_name_tra);
unlink(output_file_name_cis);

## Results:

cat('Analysis done in: ', me$time.in.sec, ' seconds', '\n');
cat('Detected local eQTLs:', '\n');

show(me$cis$eqtls)
cat('Detected distant eQTLs:', '\n');
show(me$trans$eqtls)

trans_eQTL <- me$trans$eqtls

trans_eQTL2 <- trans_eQTL %>%
  filter(FDR < 0.05)

# Get eQTLs and eGenes
# trans_eQTL <- read.table("/project/umb_brook_moyers/argo_network/trans_eQTL.txt",sep="\t",
#                                      header=T,row.names=NULL, fill=TRUE) 

# Get eGenes and eQTLs
eGenes <- data.frame(unique(trans_eQTL2$gene))
# write.table(eGenes, file = "Helianthus_argophyllus project/eGenes.txt")

eQTLs <- data.frame(unique(trans_eQTL2$snps))
# write.table(eQTLs, file = "Helianthus_argophyllus project/eQTLs.txt")

# Extract eQTL and eGenes positions
eQTL_positions <- snpspos %>%
  filter(SNP %in% trans_eQTL2$snps) %>%
  distinct()

eQTL_positions$SNP <- NULL

write.table(eQTL_positions, file = "eQTL_position.txt", sep = "\t", quote = F, row.names = F)

eGenes_positions <- genepos %>%
  filter(Geneid %in% trans_eQTL2$gene) %>%
  distinct() 

eGenes_positions$Geneid <- NULL

write.table(eGenes_positions, file = "eGenes_position.txt", sep = "\t", quote = F, row.names = F)

eGenes <- data.frame(trans_eQTL2$gene) 

eGenes <- eGenes %>%
  separate(trans_eQTL2.gene, into = c("Gene_ID", "ID", NA)) %>%
  distinct()

eGenes$Gene_ID <- NULL

# Genes with sufficient expression
suff_gene <- data.frame(row.names(expr))

suff_gene <- suff_gene %>%
  separate(row.names.expr., into = c("NA", "Gene_ID"))

length(unique(suff_gene$Gene_ID)) # 42638


# Genes with significant eQTL associations = 3806
eGenes$Gene_ID <- NULL
write.table(eGenes, file = "eGenes.txt", sep = "\t", quote = F, row.names = F)


```


